### PR TITLE
[Onboarding KPIs] Add env variables for APM instrumentation

### PR DESCRIPTION
**Pull request by liliyadd**
### What does this PR do?
Inject env variabes into application containers to be used for Onboarding KPIs.

### Motivation
https://docs.google.com/document/d/14vsrCbnAKnXmJAkacX9I6jKPGKmxsq0PKUb3dfiZpWE/edit

For APM Onboarding we want to build time-to-value telemetry. For that, we need to pass down to the tracing libraries information about injection type, injection time and unique injection id. This will help to compute the time between users installing Cluster Agent with APM enabled to the time they see the first traces.

https://datadoghq.atlassian.net/browse/APMON-366

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by Kyle-Verhoog — APPROVED** [review]

File: pkg/clusteragent/admission/mutate/config.go [hunk 1]
```diff
@@ -118,6 +142,23 @@ func injectConfig(pod *corev1.Pod, _ string, _ dynamic.Interface) error {
...
 	injectedEntity = injectEnv(pod, ddEntityIDEnvVar)
 
+	// Inject env variables used for Onboarding KPIs propagation
+	if isApmInstrumentationEnabled(pod.Namespace) {
+		// if Single Step Instrumentation is enabled, inject DD_INSTRUMENTATION_INSTALL_TYPE:k8s_single_step
+		_ = injectEnv(pod, singleStepInstrumentationInstallTypeEnvVar)
+	} else {
+		// if local library injection is enabled, inject DD_INSTRUMENTATION_INSTALL_TYPE:k8s_lib_injection
+		_ = injectEnv(pod, localLibraryInstrumentationInstallTypeEnvVar)
+	}
```
Comment (clamoriniere) [inline pkg/clusteragent/admission/mutate/config.go:152]: I think we are missing an install type, when the user already have their application instrumented with APM and they just need to inject the configuration.
 
In that cas `isApmInstrumentationEnabled` might return false, but it is not an lib injection nether...

Comment (liliyadd) [reply pkg/clusteragent/admission/mutate/config.go:152]: That's a good point. In this case we do need to check pod annotations - if library annotation is specified, then it's the case of local library injection, otherwise it's manual instrumentation. Does it make sense?

Comment (Kyle-Verhoog) [reply pkg/clusteragent/admission/mutate/config.go:152]: > I think we are missing an install type, when the user already have their application instrumented with APM and they just need to inject the configuration.

This is expected. In this case, the trace agent will apply a "manual" instrumentation type when none is provided previously.

File: pkg/clusteragent/admission/mutate/config.go [hunk 2]
```diff
@@ -118,6 +142,23 @@ func injectConfig(pod *corev1.Pod, _ string, _ dynamic.Interface) error {
...
+	// inject DD_INSTRUMENTATION_INSTALL_TIME with current Unix time
+	_ = injectEnv(pod, instrumentationInstallTimeEnvVar)
+	// inject DD_INSTRUMENTATION_INSTALL_ID with UUID created during the Agent install time
+	instrumentationInstallIDEnvVar := corev1.EnvVar{
+		Name:  instrumentationInstallIDEnvVarName,
+		Value: os.Getenv(instrumentationInstallIDEnvVarName),
```
Comment (yshapiro-57) [inline pkg/clusteragent/admission/mutate/config.go:158]: Just confirming that this environment variable is going to be set with changes to the Helm chart that will be a part of a different PR?

**Review by yshapiro-57 — APPROVED** [review]
With the disclaimer that I have limited knowledge of the cluster agent, this PR looks completely in line with what we discussed in our Zoom meetings this week.

File: pkg/clusteragent/admission/mutate/config.go (hunk 1 above)
Comment (clamoriniere) [reply pkg/clusteragent/admission/mutate/config.go:152]: but this PR change the code of an admission controller that can be use in standalone. 
how the trace-agent will understand it?

Comment (Kyle-Verhoog) [reply pkg/clusteragent/admission/mutate/config.go:152]: So how this works is that the Cluster Agent injects these environment variables in which the APM libraries will read and attach to traces. The Trace Agent will read the install_time and compute the time-to-first-trace and tag it by the install type.

Comment (clamoriniere) [reply pkg/clusteragent/admission/mutate/config.go:152]: I think I know this part.
but how the trace-agent make a different between:
* library added by the user in their container image
* library added by the admission controller?

File: pkg/clusteragent/admission/mutate/config.go [hunk 3]
```diff
@@ -76,6 +85,21 @@ var (
...
+	}
+
+	localLibraryInstrumentationInstallTypeEnvVar = corev1.EnvVar{
+		Name:  instrumentationInstallTypeEnvVarName,
+		Value: localLibraryInstrumentationInstallType,
+	}
+
+	instrumentationInstallTimeEnvVar = corev1.EnvVar{
+		Name:  instrumentationInstallTimeEnvVarName,
+		Value: strconv.FormatInt(time.Now().Unix(), 10),
```
Comment (clamoriniere) [inline pkg/clusteragent/admission/mutate/config.go:101]: setting the value with `time.Now()` here will always return the time of the cluster-agent start. 
I don't think it is was is expected. I suppose we want the timestamp when the envvar is injected.

Comment (liliyadd) [reply pkg/clusteragent/admission/mutate/config.go:101]: I don't know if it makes huge difference. We talked about DD_INSTRUMENTATION_INSTALL_TIME as the time when installer was started. In K8s context the installer is cluster-agent, so I think it's OK to use cluster agent start time. @Kyle-Verhoog what do you think?

Comment (Kyle-Verhoog) [reply pkg/clusteragent/admission/mutate/config.go:101]: We want to measure the time between intent to install APM and when they receive the first trace. To approximate the intent we use the signal of installation. So we're aiming to generate the timestamp in the helm chart/operator to pass here, if that isn't available for some reason then the startup of the Cluster Agent is the next best approximation.

Comment (clamoriniere) [reply pkg/clusteragent/admission/mutate/config.go:101]: ok it might be difficult to implement it in the helm chart too, because each time the chart will be updated the timestamp might change and so force a redeploy of the cluster-agent, which is not what we want. We only want to trigger a rollout if a configuration change.

File: pkg/clusteragent/admission/mutate/config.go [hunk 4]
```diff
@@ -27,10 +29,13 @@ import (
...
-	ddEntityIDEnvVarName   = "DD_ENTITY_ID"
-	traceURLEnvVarName     = "DD_TRACE_AGENT_URL"
-	dogstatsdURLEnvVarName = "DD_DOGSTATSD_URL"
+	agentHostEnvVarName                  = "DD_AGENT_HOST"
+	ddEntityIDEnvVarName                 = "DD_ENTITY_ID"
+	traceURLEnvVarName                   = "DD_TRACE_AGENT_URL"
+	dogstatsdURLEnvVarName               = "DD_DOGSTATSD_URL"
+	instrumentationInstallTypeEnvVarName = "DD_INSTRUMENTATION_INSTALL_TYPE"
+	instrumentationInstallTimeEnvVarName = "DD_INSTRUMENTATION_INSTALL_TIME"
+	instrumentationInstallIDEnvVarName   = "DD_INSTRUMENTATION_INSTALL_ID"
```
Comment (clamoriniere) [inline pkg/clusteragent/admission/mutate/config.go:38]: can we move this code in the `auto_instrumentation.go` file?

File: pkg/clusteragent/admission/mutate/auto_instrumentation.go [hunk 5]
```diff
@@ -134,10 +156,38 @@ func injectAutoInstrumentation(pod *corev1.Pod, _ string, _ dynamic.Interface) e
...
+	}
 
 	return injectAutoInstruConfig(pod, libsToInject, autoDetected)
 }
 
+func injectApmTelemetryConfig(pod *corev1.Pod) {
+	// inject DD_INSTRUMENTATION_INSTALL_TIME with current Unix time
+	instrumentationInstallTime := os.Getenv(instrumentationInstallTimeEnvVarName)
+	if instrumentationInstallTime == "" {
+		instrumentationInstallTime = strconv.FormatInt(time.Now().Unix(), 10)
```
Comment (Kyle-Verhoog) [inline pkg/clusteragent/admission/mutate/auto_instrumentation.go:175]: This would set the install time to the time that the library is injected, this value should be computed when the Cluster Agent starts which is closer to the intent of using APM.

**Review by clamoriniere — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]
